core/src/main/scala/cats/data/Streaming.scala

```diff
@@ -0,0 +1,947 @@
+   * diagonalization method may read ahead more than is strictly
+   * necessary.
+   */
+  def product[B](rhs: Streaming[B]): Streaming[(A, B)] = {
+    def loop(i: Int): Streaming[(A, B)] = {
+      val xs = lhs.take(i + 1).asInstanceOf[Streaming[AnyRef]].toArray
+      val ys = rhs.take(i + 1).asInstanceOf[Streaming[AnyRef]].toArray
+      def build(j: Int): Streaming[(A, B)] =
+        if (j > i) Empty() else {
+          val k = i - j
+          if (j >= xs.length || k >= ys.length) build(j + 1) else {
+            val tpl = (xs(j).asInstanceOf[A], ys(k).asInstanceOf[B])
+            Cons(tpl, Always(build(j + 1)))
+          }
+        }
+      if (i > xs.length + ys.length - 2) Empty() else {
+        build(0) concat Always(loop(i + 1))
+      }
+    }
+    Wait(Always(loop(0)))
+  }
+
+  /**
+   * Return true if some element of the stream satisfies the
+   * predicate, false otherwise.
+   */
+  def exists(f: A => Boolean): Boolean = {
+    @tailrec def unroll(s: Streaming[A]): Boolean =
+      s match {
+        case Empty() => false
+        case Wait(lt) => unroll(lt.value)
+        case Cons(a, lt) => if (f(a)) true else unroll(lt.value)
+      }
+    unroll(this)
+  }
+
+  /**
+   * Return true if every element of the stream satisfies the
+   * predicate, false otherwise.
+   */
+  def forall(f: A => Boolean): Boolean = {
+    @tailrec def unroll(s: Streaming[A]): Boolean =
+      s match {
+        case Empty() => true
+        case Wait(lt) => unroll(lt.value)
+        case Cons(a, lt) => if (f(a)) unroll(lt.value) else false
+      }
+    unroll(this)
+  }
+
+  /**
+   * Return a stream consisting only of the first `n` elements of this
+   * stream.
+   *
+   * If the current stream has `n` or fewer elements, the entire
+   * stream will be returned.
+   */
+  def take(n: Int): Streaming[A] =
+    if (n <= 0) Empty() else this match {
+      case Empty() => Empty()
+      case Wait(lt) => Wait(lt.map(_.take(n)))
+      case Cons(a, lt) => Cons(a, lt.map(_.take(n - 1)))
+    }
+
+  /**
+   * Return a stream consisting of all but the first `n` elements of
+   * this stream.
+   *
+   * If the current stream has `n` or fewer elements, an empty stream
+   * will be returned.
+   */
+  def drop(n: Int): Streaming[A] =
+    if (n <= 0) this else this match {
+      case Empty() => Empty()
+      case Wait(lt) => Wait(lt.map(_.drop(n)))
+      case Cons(a, lt) => Wait(lt.map(_.drop(n - 1)))
+    }
+
+  /**
+   * From the beginning of this stream, create a new stream which
+   * takes only those elements that fulfill the predicate `f`. Once an
+   * element is found which does not fulfill the predicate, no further
+   * elements will be returned.
+   *
+   * If all elements satisfy `f`, the current stream will be returned.
+   * If no elements satisfy `f`, an empty stream will be returned.
+   *
+   * For example:
+   *
+   *   Stream(1, 2, 3, 4, 5, 6, 7).takeWhile(n => n != 4)
```

Comment:
There are still some references to `Stream` instead of `Streaming` in the ScalaDocs (see also `dropWhile` and the `syntax` object). Some day we should start using [sbt-doctest](https://github.com/tkawachi/sbt-doctest). :-)